### PR TITLE
Add option in box.cfg that allows to select allocator for memtx tuples

### LINKED ISSUE #5419: Choose an allocator for memtx in box.cfg
Slab allocator, which is used for allocation tuples, has a certain disadvantage - it tends to unresolvable  fragmentation on certain workloads (size migration).

Provide an option in box.cfg that allows to choose some other allocator for memtx tuples. At least it must be simple allocator like hpall (there must a branch). At most - log structurea allocator (like lsall, also on the same branch).

There's also an idea to make that option per space, but I'm not sure.

### PR DESCRIPTION
**Pull request by mechanik20051988**
Slab allocator, which is used for allocation tuples, has a certain
disadvantage - it tends to unresolvable fragmentation on certain
workloads (size migration). New option allows to select
the appropriate allocator if necessary.

@TarantoolBot document
Title: Add new 'allocator' option
Add new 'allocator' option which allows to select
the appropriate allocator for memtx tuples if necessary.
Use box.cfg{allocator="small"} or no option to use default
small allocator, use box.cfg{allocator="system"} to use
libc malloc.

Closes #5419

### CODE REVIEW

**Comment by Gerold103** [comment]
Can you provide more details about performance impact, and elaborate the options in the doc request? Their values, meaning.
Also, tuple_new is already virtual. Why can't you reuse this virtuality instead of introducing +2 virtual calls for each tuple to allocate/delete it?

**Comment by mechanik20051988** [comment]
> Can you provide more details about performance impact, and elaborate the options in the doc request? Their values, meaning.
> Also, tuple_new is already virtual. Why can't you reuse this virtuality instead of introducing +2 virtual calls for each tuple to allocate/delete it?
I 


> Can you provide more details about performance impact, and elaborate the options in the doc request? Their values, meaning.
> Also, tuple_new is already virtual. Why can't you reuse this virtuality instead of introducing +2 virtual calls for each tuple to allocate/delete it?

I write perfot

**Comment by mechanik20051988** [comment]
> Can you provide more details about performance impact, and elaborate the options in the doc request? Their values, meaning.
> Also, tuple_new is already virtual. Why can't you reuse this virtuality instead of introducing +2 virtual calls for each tuple to allocate/delete it?

Sorry closed inadvertently!! I write performance test and see 2% performance lost because of my patch, you idea about reuse tuple_new is good, i try it now to see performance

**Comment by mary3000** [comment]
Minor: subject lines in commit message should be no more than 50 characters

**Comment by mary3000** [comment]
> I write performance test and see 2% performance lost because of my patch, you idea about reuse tuple_new is good, i try it now to see performance
Could you please add a comment to a pr with final perf lost (if any), if you tried this idea?

**Comment by mary3000** [comment]
Also, I think it's reasonable to add a simple test to check just functionality, without performance (plus /test is integrated into CI, whereas /perf - currently not)

**Comment by mary3000** [comment]
Please ensure that each file fits 80 char line limit

File: src/box/memtx_engine.c [hunk 1]
```diff
@@ -1061,6 +1244,42 @@ memtx_engine_new(const char *snap_dirname, bool force_recovery,
...
+		memtx_tuple_format_vtab.tuple_delete = system_memtx_tuple_delete;
+		memtx_tuple_format_vtab.tuple_new = system_memtx_tuple_new;
+		memtx_tuple_format_vtab.tuple_chunk_delete = system_memtx_tuple_chunk_delete;
+		memtx_tuple_format_vtab.tuple_chunk_new = system_memtx_tuple_chunk_new;
+		memtx_allocator_init = system_allocator_init;
+		memtx_allocator_destroy = system_allocator_destroy;
+		memtx_allocator_enter_delayed_free_mode = system_allocator_enter_delayed_free_mode;
+		memtx_allocator_leave_delayed_free_mode = system_allocator_leave_delayed_free_mode;
+		memtx_allocator_stats = system_allocator_stats;
+		memtx_slab_cache_check = system_slab_cache_check;
```
Comment (mary3000) [inline src/box/memtx_engine.c:1276]: If we're going to introduce other allocators later, there would be even more such lengthy assignments. I suggest adding a macro for it.

**Comment by mary3000** [comment]
Could you please summarize the purpose of the performance test? Is it going to be integrated into CI, will be there any checks that performance will not degrade for example? Why this workload? Compare of system vs small allocator on this perf test?

**Comment by mary3000** [comment]
Also, as I know malloc can be replaced with another allocator. If this behavior is accepted, then I guess it should be stated in docs and maybe even tested?

**Comment by mechanik20051988** [comment]
> > I write performance test and see 2% performance lost because of my patch, you idea about reuse tuple_new is good, i try it now to see performance
> 
> Could you please add a comment to a pr with final perf lost (if any), if you tried this idea?
In this realisation, there was no perf lost at all. It was in previous commit (deleted by force push), where i use allocator struct as vtab with methods for (allocate/free memory)

**Comment by mechanik20051988** [comment]
> Could you please summarize the purpose of the performance test? Is it going to be integrated into CI, will be there any checks that performance will not degrade for example? Why this workload? Compare of system vs small allocator on this perf test? About workload: we tested all main operations which used memory allocator(insert/update/delete), I chose the parameters so that the test took no more than 5 minutes on my computer

The perf test is currently being run manually. You cat run it without args to see perf for default small allocator(also you can copy it to master branch and see perf in master, compare it and see no perf lost)

File: src/box/system_allocator.h [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		size_t used_bytes =  pm_atomic_load(&alloc->used_bytes);
+		if (small_align(used_bytes, QUOTA_UNIT_SIZE) < small_align(used_bytes + s, QUOTA_UNIT_SIZE))
+			units++;
+		if (units > 0) {
+			if (quota_use(alloc->quota, units * QUOTA_UNIT_SIZE) < 0) {
+				free(ptr);
+				return NULL;
+			}
+		}
+		if (pm_atomic_compare_exchange_strong(&alloc->used_bytes, &used_bytes, used_bytes + size))
```
[the remark this answers is not in the document]
Comment (mechanik20051988) [reply src/box/system_allocator.h]: I write according to src/lib/small/small/quota.h there we used strong version

**Comment by mechanik20051988** [comment]
> Also, as I know malloc can be replaced with another allocator. If this behavior is accepted, then I guess it should be stated in docs and maybe even tested?

This task is a first step. The next step is to add good heap allocator (better then malloc). This step only provide user to choose malloc.

**Comment by mary3000** [comment]
LGTM, passing to @alyapunov

File: src/box/allocator.h [hunk 3]
```diff
@@ -0,0 +1,153 @@
+#ifndef TARANTOOL_BOX_ALLOCATOR_H_INCLUDED
+#define TARANTOOL_BOX_ALLOCATOR_H_INCLUDED
```
Comment (Gerold103) [inline src/box/allocator.h:2]: Please, use `#pragma once`. This is a new rule, which is not added anywhere yet, sorry. Kirill approved it and was supposed to update the documentation.

File: src/box/allocator.h [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+ */
+typedef void (*global_slab_cache_check)(struct memtx_engine *memtx);
+extern global_slab_cache_check memtx_slab_cache_check;
+
+#if defined(__cplusplus)
+} /* extern "C" */
+#endif /* defined(__cplusplus) */
+
+
+#endif /* TARANTOOL_BOX_ALLOCATOR_H_INCLUDED */
```
Comment (Gerold103) [inline src/box/allocator.h:153]: Git warns you about missing last empty line. Please, don't ignore.

File: src/box/memtx_engine.c [hunk 5]
```diff
@@ -50,10 +50,16 @@
...
 #include "raft.h"
+#include "allocator.h"
 
 /* sync snapshot every 16MB */
 #define SNAP_SYNC_INTERVAL	(1 << 24)
 
+static global_enter_delayed_free_mode memtx_allocator_enter_delayed_free_mode;
+static global_leave_delayed_free_mode memtx_allocator_leave_delayed_free_mode;
+static global_allocator_destroy memtx_allocator_destroy;
+static global_allocator_init memtx_allocator_init;
```
Comment (Gerold103) [inline src/box/memtx_engine.c:61]: Why are these global variables? We have memtx_engine to all memtx-specific things. These clearly are memtx-specific.

File: src/box/memtx_engine.c [hunk 6]
```diff
@@ -1049,10 +1053,220 @@ memtx_engine_gc_f(va_list va)
...
+void
+memtx_enter_delayed_free_mode(struct memtx_engine *memtx)
+{
+	memtx->snapshot_version++;
+	if (memtx->delayed_free_mode++ == 0)
+		memtx_allocator_enter_delayed_free_mode(memtx);
+}
+
+void
+memtx_leave_delayed_free_mode(struct memtx_engine *memtx)
```
Comment (Gerold103) [inline src/box/memtx_engine.c:1097]: These changes look totally unrelated to the an allocator type. Please, extract such non-functional things to separate commits.

File: src/box/memtx_engine.c [hunk 7]
```diff
@@ -1049,10 +1053,220 @@ memtx_engine_gc_f(va_list va)
...
+	tuple->data_offset = data_offset;					\
+	tuple->is_dirty = false;						\
+	char *raw = (char *) tuple + tuple->data_offset;			\
+	field_map_build(&builder, raw - field_map_size);			\
+	memcpy(raw, data, tuple_len);						\
+	say_debug("%s(%zu) = %p", __func__, tuple_len, memtx_tuple);		\
+end:										\
+	region_truncate(region, region_svp);					\
+	return tuple;								\
+}
```
Comment (Gerold103) [inline src/box/memtx_engine.c:1173]: Instead of this macro-hell (more than 100 lines of macros!!!) I would suggest to turn this file into .cpp, and try templates maybe. If possible.

File: src/box/memtx_engine.c [hunk 8]
```diff
@@ -1138,174 +1363,6 @@ memtx_engine_new(const char *snap_dirname, bool force_recovery,
 	return NULL;
```
Comment (Gerold103) [inline src/box/memtx_engine.c:1363]: Allocator leaks here.

File: src/box/system_allocator.h [hunk 9]
```diff
@@ -0,0 +1,181 @@
+#ifndef TARANTOOL_BOX_SYSTEM_ALLOCATOR_H_INCLUDED
+#define TARANTOOL_BOX_SYSTEM_ALLOCATOR_H_INCLUDED
+/*
+ * Copyright 2010-2016, Tarantool AUTHORS, please see AUTHORS file.
```
Comment (Gerold103) [inline src/box/system_allocator.h:4]: I know 2020 is bad, but we can't simply return to 2016 like that.

File: src/box/system_allocator.h [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+portable_malloc_usable_size(void *p)
+{
+	return malloc_usable_size(p);
+}
+#elif defined(__linux__)
+#include <malloc.h>
+static inline size_t
+portable_malloc_usable_size(void *p)
+{
+	return malloc_usable_size(p);
```
Comment (Gerold103) [inline src/box/system_allocator.h:60]: You may want to consider using CMake command `check_symbol_exists` instead of dealing with the operating system macros. And even if you would need that, we have TARGET_OS_LINUX, TARGET_OS_DARWIN, and so on.

File: src/box/system_allocator.h [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+static inline void
+system_alloc_create(struct system_alloc *alloc, struct quota *quota,
+	uint32_t objsize_min, float alloc_factor)
+{
+	alloc->used_bytes = 0;
+	alloc->quota = quota;
+	(void)objsize_min;
+	(void)alloc_factor;
```
Comment (Gerold103) [inline src/box/system_allocator.h:119]: Why are these parameters passed here, if they are not used?

File: src/box/system_allocator.h [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+static inline void
+system_alloc_destroy(MAYBE_UNUSED struct system_alloc *alloc)
+{
+
```
Comment (Gerold103) [inline src/box/system_allocator.h:125]: So if sysfree() will happen later, it will crash, because alloc may be already deleted? I suggest you to add a panic here in case not everything is freed, maybe.

File: src/box/system_allocator.h [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+		units++;
+	if (units > 0)
+		quota_release(alloc->quota, units * QUOTA_UNIT_SIZE);
+	free(ptr);
+}
+
+static inline void
+sysfree_delayed(struct system_alloc *alloc, void *ptr, MAYBE_UNUSED unsigned bytes)
+{
+	return sysfree(alloc, ptr, bytes);
```
Comment (Gerold103) [inline src/box/system_allocator.h:145]: I think I already asked it somewhere, but how does it work the snapshots? Delayed was supposed not to do anything until snapshot ends. And here you simply free it. I suggest you to add something like`memset('#')` for the freed memory to see a crash or some other fail earlier, if the memory gets used after deletion.
I see the snapshots are not even tested for the new allocator.

**Review by Gerold103 — COMMENTED** [review]
I took a quick glance at it without diving into details. Hopefully these comments are helpful.

**Comment by mechanik20051988** [comment]
> I took a quick glance at it without diving into details. Hopefully these comments are helpful.

Thank you for review. I will correct all and write!

File: src/box/system_allocator.h (hunk 10 above)
Comment (mechanik20051988) [reply src/box/system_allocator.h:59]: Fix using check_symbol_exists

File: src/box/system_allocator.h (hunk 12 above)
Comment (mechanik20051988) [reply src/box/system_allocator.h]: As i see in small_alloc_destroy we not check that all memory was free, before we destroy allocator. I add some debug info in memtx_engine_shutdown function (i call small_stats and print result), where we destroy allocator and see that not all memory was freed (in master branch also). So i add init flag which set in system_allocator_init and reset in system_allocator_destroy, also i add asserts in all allocator functions to check that flag is set (i check that allocator not used before init and after destroy). But the full release of memory remains on the conscience of the allocator user. How do you think its ok?

**Comment by mechanik20051988** [comment]
I also done this work without macros in mechanik20051988/gh-5419-choose-allocator-for-memtx-cpp.

File: src/box/system_allocator.h (hunk 11 above)
Comment (mechanik20051988) [reply src/box/system_allocator.h]: Fixed

File: src/box/system_allocator.h (hunk 13 above)
Comment (mechanik20051988) [reply src/box/system_allocator.h]: Fixed

File: src/box/system_allocator.h (hunk 9 above)
Comment (mechanik20051988) [reply src/box/system_allocator.h]: Fixed

File: src/box/memtx_engine.c (hunk 7 above)
Comment (mechanik20051988) [reply src/box/memtx_engine.c]: I significantly reduced the number of macros and encapsulated them in the allocator.h. Also i done this work without macros in mechanik20051988/gh-5419-choose-allocator-for-memtx-cpp using templates.

File: src/box/memtx_engine.c (hunk 6 above)
Comment (mechanik20051988) [reply src/box/memtx_engine.c]: Fixed

File: src/box/memtx_engine.c (hunk 5 above)
Comment (mechanik20051988) [reply src/box/memtx_engine.c]: Fixed

File: src/box/allocator.h (hunk 4 above)
Comment (mechanik20051988) [reply src/box/allocator.h]: Fixed

File: src/box/allocator.h (hunk 3 above)
Comment (mechanik20051988) [reply src/box/allocator.h]: Fixed
